src/Microsoft.AspNet.Hosting/Internal/HostingApplication.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Hosting.Server;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    public class HostingApplication : IHttpApplication<HostingApplication.Context>
+    {
+        private readonly IServiceProvider _applicationServices;
+        private readonly RequestDelegate _application;
+        private readonly ILogger _logger;
+        private readonly DiagnosticSource _diagnosticSource;
+        private readonly IHttpContextFactory _httpContextFactory;
+
+        public HostingApplication(
+            IServiceProvider applicationServices,
+            RequestDelegate application,
+            ILogger logger,
+            DiagnosticSource diagnosticSource,
+            IHttpContextFactory httpContextFactory)
+        {
+            _applicationServices = applicationServices;
+            _application = application;
+            _logger = logger;
+            _diagnosticSource = diagnosticSource;
+            _httpContextFactory = httpContextFactory;
+        }
+
+        public Context CreateContext(IFeatureCollection contextFeatures)
+        {
+            var httpContext = _httpContextFactory.Create(contextFeatures);
+
+            var scope = _logger.RequestScope(httpContext);
+            _logger.RequestStarting(httpContext);
+            if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Hosting.BeginRequest"))
+            {
+                _diagnosticSource.Write("Microsoft.AspNet.Hosting.BeginRequest", new { httpContext });
+            }
+
+            return new Context
+            {
+                HttpContext = httpContext,
+                Scope = scope,
+                TickCount = Environment.TickCount,
+            };
+        }
+
+        public void DisposeContext(Context context, Exception exception = null)
+        {
+            var httpContext = context.HttpContext;
+            var elapsed = new TimeSpan(Environment.TickCount - context.TickCount);
```

Comment:
Isn't this a bit fragile? IIRC, `Environment.TickCount` resets after ~25 days of uptime. If it happens in the middle of a request, the result could be really strange :smile: